### PR TITLE
posix: pthread: return value not set by pthread_join()

### LINKED ISSUE #60960: posix: pthread: return value is not set in pthread_join()
**Describe the bug**
POSIX thread functions have the signature `void *func(void *arg)`.

The join function has the signature `int pthread_join(pthread_t thread, void **ret)`.

Where `ret` (if non-NULL) is set to the return value of the thread function.

Currently, the return value is not being set.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? any
 - What have you tried to diagnose or workaround this issue? test
 - Is this a regression? Yes (not sure when it was introduced)

**To Reproduce**
1. add testcase below
2. `west build -p auto -b qemu_x86 -t run tests/posix/common`
3. See error

**Expected behavior**
Testsuite should pass. Return value of thread should be available. Implementation should be compliant.

**Impact**
Likely impacts anyone using the POSIX API for threading.

**Logs and console output**
```
START - test_pthread_return_val

    Assertion failed at WEST_TOPDIR/zephyr/tests/posix/common/src/pthread.c:834: posix_apis_test_pthread_return_val: ret not equal to (void *)0x42427373

 FAIL - test_pthread_return_val in 0.002 seconds
===================================================================
```

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows) Linux
 - Toolchain (e.g Zephyr SDK, ...):  Zephyr SDK 0.16.1
 - Commit SHA or Version used: 9dc531be481a2e67da5c2875339b4386e93e823c

**Additional context**
https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html

Testcase:
```cpp
static void *non_null_retval(void *arg)
{
	ARG_UNUSED(arg);

	return (void *)0x42427373;
}

ZTEST(posix_apis, test_pthread_return_val)
{
	pthread_t pth;
	void *ret = NULL;
	pthread_attr_t attr;

	zassert_ok(pthread_attr_init(&attr));
	zassert_ok(pthread_attr_setstack(&attr, &stack_e[0][0], STACKS));

	zassert_ok(pthread_create(&pth, &attr, non_null_retval, NULL));
	zassert_ok(pthread_join(pth, &ret));
	zassert_equal(ret, (void *)0x42427373);
}
```

### PR DESCRIPTION
**Pull request by cfriedt**
Ensure that the thread return value is set by `pthread_join()` when `status` is non-NULL.
 
 Fixes #60960

### CODE REVIEW

**Review by ycsin — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

File: tests/posix/common/src/pthread.c [hunk 1]
```diff
@@ -812,3 +812,24 @@ ZTEST(posix_apis, test_pthread_dynamic_stacks)
 	zassert_ok(pthread_join(th, NULL));
 	zassert_equal(0xB105F00D, x);
 }
+
```
Comment (keith-packard) [inline tests/posix/common/src/pthread.c:815]: Might use a #define value here to make it a bit more obvious what's being tested?

File: lib/posix/pthread.c [hunk 2]
```diff
@@ -656,6 +656,12 @@ int pthread_join(pthread_t pthread, void **status)
...
 		__ASSERT_NO_MSG(err == 0);
+
+		if (status != NULL) {
+			*status = t->retval;
+		}
+
+		posix_thread_recycle();
 	}
 
 	return 0;
```
Comment (keith-packard) [inline lib/posix/pthread.c:667]: Looks like this should be `return ret` so that you get `ESRCH` or `EINVAL` when there's an error?

**Review by keith-packard — DISMISSED** [review]
New change looks fine, but I think I found an existing bug.

File: lib/posix/pthread.c (hunk 2 above)
Comment (fabiobaltieri) [reply lib/posix/pthread.c:667]: Looks like https://github.com/zephyrproject-rtos/zephyr/pull/60964/files covers that. How about changing 641 and 644 to a pair of `k_spin_unlock(); return E...;` and then dropping the `if (ret == 0) {` and just handle the success case inline?

**Comment by cfriedt** [comment]
Hehe..  sorry for the multiple PRs, but I thought each regression deserved its own issue .. I guess I could have put them together.

File: lib/posix/pthread.c (hunk 2 above)
Comment (cfriedt) [reply lib/posix/pthread.c:667]: Exactly - fixed in a separate PR

**Comment by cfriedt** [comment]
* resolved merge conflict

**Comment by fabiobaltieri** [comment]
@cfriedt how about the refactoring I mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/60961#discussion_r1278591265? I think it may make that function a bit less tangled.

File: lib/posix/pthread.c (hunk 2 above)
Comment (cfriedt) [reply lib/posix/pthread.c]: > How about changing 641 and 644 to a pair of k_spin_unlock(); return E...;, inline the else case up there and then dropping the if (ret == 0) { and just handle the success case inline?

That sounds reasonable, let me give it another spin.

Comment (cfriedt) [reply lib/posix/pthread.c:671]: @fabiobaltieri - how does this look?

File: tests/posix/common/src/pthread.c (hunk 1 above)
Comment (cfriedt) [reply tests/posix/common/src/pthread.c:818]: Fixed!

**Comment by cfriedt** [comment]
> @cfriedt how about the refactoring I mentioned in [#60961 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/60961#discussion_r1278591265)? I think it may make that function a bit less tangled.

@fabiobaltieri  - does it look ok now?

File: lib/posix/pthread.c (hunk 2 above)
Comment (fabiobaltieri) [reply lib/posix/pthread.c:671]: fancy...

**Review by fabiobaltieri — DISMISSED** [review]

**Review by keith-packard — DISMISSED** [review]
Looks like compliance is whinging about a comment format. Otherwise, this looks simply fabulous.

**Comment by cfriedt** [comment]
> Looks like compliance is whinging about a comment format. Otherwise, this looks simply fabulous.

@fabiobaltieri @keith-packard - fixed!

File: lib/posix/pthread.c [hunk 3]
```diff
@@ -636,29 +635,42 @@ int pthread_join(pthread_t pthread, void **status)
...
-	} else if (t->qid == POSIX_THREAD_READY_Q) {
-		/* marginal chance thread has moved to ready_q between to_posix_thread() and here */
-		ret = ESRCH;
-	} else {
+	K_SPINLOCK(&pthread_pool_lock)
+	{
+		if (t->detachstate != PTHREAD_CREATE_JOINABLE) {
+			ret = EINVAL;
+			K_SPINLOCK_BREAK;
+		} else if (t->qid == POSIX_THREAD_READY_Q) {
```
Comment (keith-packard) [inline lib/posix/pthread.c:643]: minor comment here -- you're using `K_SPINLOCK_BREAK` in the `if` clause above, so you don't actually need the `else` here. Any idea what conventions with `K_SPINLOCK_BREAK` are these days? And, thanks for exposing us to K_SPINLOCK. That's a sweet macro.

**Review by keith-packard — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

File: lib/posix/pthread.c (hunk 3 above)
Comment (cfriedt) [reply lib/posix/pthread.c]: Fixed!

**Review by fabiobaltieri — APPROVED** [review]

**Review by keith-packard — APPROVED** [review]

**Pull request merged**